### PR TITLE
Drop melted users

### PR DESCRIPTION
**Pull request by jrauschenberg**
#### What's this PR do?
It checks for users who have been marked by trainers in Canvas as "deactivated," finds them in WiW, and deletes them and their shifts.
#### Where should the reviewer start?
jobs/scheduling/dropMeltedUsers
#### How should this be manually tested?
npm test, and/or:
1. Create a fake Canvas user.
2. Enroll the user in a course.
3. Select "deactivate user" for your user in the list of students in the course.
4. Create a user with the same email address in WiW.
5. Sign that user up for a shift or two in WiW.
6. Run dropMeltedUsers with node (including the appropriate CONFIG and KEYS variables as needed to help it run). Alternatively, use runJobs, but only set it to run dropMeltedUsers.
7. Make sure the user's WiW account no longer exists and their shifts no longer belong to them.
#### What are the relevant tickets?
[INT-130](https://admin.crisistextline.org/jira/secure/RapidBoard.jspa?rapidView=5&view=detail&selectedIssue=INT-130), [INT-131](https://admin.crisistextline.org/jira/secure/RapidBoard.jspa?rapidView=5&view=detail&selectedIssue=INT-131), [INT-132](https://admin.crisistextline.org/jira/secure/RapidBoard.jspa?rapidView=5&view=detail&selectedIssue=INT-132)

### CODE REVIEW

File: jobs/scheduling/dropMeltedUsers.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  })
+  .catch(function(err){
+    CONSOLE_WITH_TIME("Finding enrollments for the course with ID " + courseID + " in Canvas failed: ", err);
+  });
+
+};
+
+deleteWiWUserAndShifts = function(canvasUser) {
+  var usersForTest;
+  WhenIWork.get('users', CONFIG.locationID.test, function(users) {
```
Comment (apackin) [inline jobs/scheduling/dropMeltedUsers.js]: Accidentally left this set to CONFIG.locationID.test ?

We're making this get request for all users to WiW for each user in the enrollments objects returned from canvas. Is there a way we could request all the users once and then keep using that same return for each user in enrollments? 

Maybe add get users as a req in the promise chain of findMeltedCanvasUsersAndDeleteThemInWiW()?

File: test/dropMeltedUsers.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+var assert = require('assert');
+var dropMeltedUsers = require('../jobs/scheduling/dropMeltedUsers.js');
+var canvasUser = { id: 1734,
```
Comment (apackin) [inline test/dropMeltedUsers.js]: Where is this used?

File: test/dropMeltedUsers.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+       sis_login_id: 'john@crisistextline.org',
+       sis_import_id: null,
+       login_id: 'john@crisistextline.org' };
+
+describe('drop melted users', function() {
+
+    it('should find all courses in Canvas', function (done) {
+      dropMeltedUsers.canvas.scrapeCanvasCourses()
+      .then(function(result) {
+        assert.equal(result[0].id, 1);
```
Comment (apackin) [inline test/dropMeltedUsers.js]: If it should find all courses in Canvas then we should assert that in the test. We're only checking that the first course is right.
On second thought that function is only an API call, I'm not sure that we need to test it.

**Comment by apackin** [comment]
I really like the modularity of this code. I'm running npm test and failing on "drop melted users should find all courses in Canvas Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test."

Are the email templates used for this PR?

**Comment by jrauschenberg** [comment]
@apackin Fixed everything according to your suggestions. I kept the Canvas course-finding test in but added a length check to make sure it's getting LOTS of courses. I also removed the email templates. I'm not getting that timeout. I think if you run it a couple more times you won't get it.

**Comment by tongxiang** [comment]
Why would one box be getting the timeout and another box *not* getting the timeout?

File: test/dropMeltedUsers.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        if (result.length >= 40) resultIsLong = true;
+        assert.equal(resultIsLong, true);
+        done();
+      });
+    });
+
+    it('should find melted users in Canvas', function (done) {
+      dropMeltedUsers.canvas.scrapeCanvasEnrollment(60, 'inactive')
+      .then(function(result) {
+        assert.equal(result[0].user_id, 1734);
```
Comment (tongxiang) [inline test/dropMeltedUsers.js]: to make these tests easier to read, definitely recommend assigning constants as variables first before putting them in `assert` statements.

File: jobs/scheduling/dropMeltedUsers.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    Authorization: KEYS.canvas.api_key,
+    'User-Agent': 'Request-Promise'
+  }
+};
+var canvas = {};
+
+new CronJob(CONFIG.time_interval.cron_twice_per_day, findMeltedCanvasUsersAndDeleteThemInWiW);
+
+//finds all canvas courses
+canvas.scrapeCanvasCourses = function() {
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: technically we're just querying an API, not scraping (which really is just converting HTML into a machine-readable format)

more correct variable name would just be `canvas.retrieveCanvasCourses`

Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: same with `canvas.scrapeCanvasEnrollment`

Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: Or even better, `canvas.scrapeCourses` and `canvas.scrapeEnrollment`

Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: Also, why are we instantiating these canvas functions here when they already exist in `updateCanvas.js`?

Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: If we're adding new functions to canvas, we can just add it to `updateCanvas.js`

File: jobs/scheduling/dropMeltedUsers.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    response = JSON.parse(response);
+    return response;
+  })
+  .catch(function(err){
+    CONSOLE_WITH_TIME("Finding enrollments for the course with ID " + courseID + " in Canvas failed: ", err);
+  });
+
+};
+
+deleteWiWUserAndShifts = function(canvasUser, WiWUsers) {
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: why are we making this global?

it looks like there's a function declaration immediately below this function expression. 

Let's stick to one, for consistency's sake. (Probably function declarations would be better.)

File: jobs/scheduling/dropMeltedUsers.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      return result;
+    })
+    .then(function(courseIDs) {
+      courseIDs.forEach(function(courseID) {
+        return canvas.scrapeCanvasEnrollment(courseID, 'inactive')
+        .then(function(enrollments) {
+          if (!enrollments) throw 'No inactive enrollments found.';
+          enrollments.forEach(function(enrollment) {
+            //take user object here and delete user's WiW shifts + account
+            deleteWiWUserAndShifts(enrollment.user, users);
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: add a `.catch` block to atch errors thrown by the `canvas.scrapeCanvasEnrollment` call

File: jobs/scheduling/dropMeltedUsers.js (hunk 6 above)
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: Can we add an error callback here?

**Comment by tongxiang** [comment]
Merging this in is also dependent on this feature which @mbogor is working on: 

https://admin.crisistextline.org/jira/browse/INT-226

File: jobs/scheduling/dropMeltedUsers.js (hunk 7 above)
Comment (jrauschenberg) [inline jobs/scheduling/dropMeltedUsers.js]: I think there is a catch block below this code. Let me know if I'm missing
something!

File: jobs/scheduling/dropMeltedUsers.js (hunk 5 above)
Comment (jrauschenberg) [inline jobs/scheduling/dropMeltedUsers.js]: Mariya and I would like to work on integrating all our Canvas functions
into one file. I'm hoping to do it soon. Can we get some time next sprint?

File: jobs/scheduling/dropMeltedUsers.js (hunk 6 above)
Comment (jrauschenberg) [inline jobs/scheduling/dropMeltedUsers.js]: The catch block includes error handling, but I think I'm missing something.
What kind of callback are you talking about?

File: jobs/scheduling/dropMeltedUsers.js (hunk 7 above)
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: this is a nested `.then` statement, so the errors within here will not be caught by the `catch` statement which listens for errors in the outside `.then` statement, I believe @jrauschenberg

Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: we want to confirm that the errors bubble up

File: jobs/scheduling/dropMeltedUsers.js (hunk 5 above)
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: Totally--wanna create an issue for that? 

I think it might still make sense to just call those functions in this PR, though--is there a reason why you'd want to do it in another PR?

Comment (jrauschenberg) [inline jobs/scheduling/dropMeltedUsers.js]: Mariya has created an issue already--check her JIRA. I can call the functions in this PR for now.

**Comment by jrauschenberg** [comment]
Hey @tongxiang I think I addressed all the issues you mentioned but let me know if I missed anything. Thanks!

File: jobs/scheduling/dropMeltedUsers.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+var WhenIWork = CONFIG.WhenIWork;
+var updateCanvas = require('./updateCanvas.js');
+var fs = require('fs');
+var options = {
+  headers: {
+    Authorization: KEYS.canvas.api_key,
+    'User-Agent': 'Request-Promise'
+  }
+};
+var canvas = {};
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: probably can take this out

File: jobs/scheduling/dropMeltedUsers.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+new CronJob(CONFIG.time_interval.cron_twice_per_day, findMeltedCanvasUsersAndDeleteThemInWiW);
+
+// //finds all canvas courses
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: why did you leave this commented out? 

probably can take this out

**Comment by jrauschenberg** [comment]
Yeah, I guess I was just paranoid about removing the old Canvas code, but it does live in another file now so I killed my baby. Good to merge?

**Comment by tongxiang** [comment]
Lolz @jrauschenberg think you notified some other person with your last comment. Will check this out again.

**Comment by jrauschenberg** [comment]
Ooops, sorry @tongxiang, and other Tong.

File: test/index.js [hunk 10]
```diff
@@ -10,14 +10,17 @@ global.CONFIG = require('../config.js');
 var recurOpenShifts = require('./recurOpenShifts.js');
 var colorize = require('./colorize.js');
 var attendanceSync = require('./attendanceSync.js');
+var colorize = require('./colorize.js');
```
Comment (tongxiang) [inline test/index.js]: why did we add this again?

File: test/dropMeltedUsers.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
+var assert = require('assert');
+var dropMeltedUsers = require('../jobs/scheduling/dropMeltedUsers.js');
+var updateCanvas = require('../jobs/scheduling/helpers/updateCanvas.js');
+var WiWUsers = require('../sample_data/sampleData').usersResponse;
+var canvasUserID = 1734;
+var WiWUserID = 7889841;
+
```
Comment (tongxiang) [inline test/dropMeltedUsers.js]: since the variables are relevant specifically to the tests, and aren't files being required, I would move the variable declarations inside the `describe` block

File: test/index.js [hunk 12]
```diff
@@ -10,14 +10,17 @@ global.CONFIG = require('../config.js');
...
 var loginUserCreation = require('./loginUserCreation.js');
 var retrieveAndRenderShiftsToDelete = require('./retrieveAndRenderShiftsToDelete.js');
 var whenIWorkUnofficialAPI = require('./whenIWorkUnofficialAPI.js');
 var timezone = require('./timezone.js');
 var deleteShiftsAndRedirect = require('./deleteShiftsAndRedirect.js');
 var renderShiftDeleteSuccess = require('./renderShiftDeleteSuccess.js');
 var mergeOpenShifts = require('./mergeOpenShifts.js');
+var notifyFirstShift = require('./notifyFirstShift.js');
```
Comment (tongxiang) [inline test/index.js]: module.js:442
    throw err;
    ^

Error: Cannot find module './notifyFirstShift.js'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/tongxiang/ctl/Elmer/test/index.js:21:24)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at /Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:219:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:216:14)
    at Mocha.run (/Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:468:10)
    at Object.<anonymous> (/Users/tongxiang/ctl/Elmer/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (node.js:348:7)
    at startup (node.js:140:9)
    at node.js:463:3
npm ERR! Test failed.  See above for more details.

**Comment by tongxiang** [comment]
module.js:442
    throw err;
    ^

Error: Cannot find module './notifyFirstShift.js'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/tongxiang/ctl/Elmer/test/index.js:21:24)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at /Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:219:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:216:14)
    at Mocha.run (/Users/tongxiang/ctl/Elmer/node_modules/mocha/lib/mocha.js:468:10)
    at Object.<anonymous> (/Users/tongxiang/ctl/Elmer/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (node.js:348:7)
    at startup (node.js:140:9)
    at node.js:463:3
npm ERR! Test failed.  See above for more details.

File: jobs/scheduling/dropMeltedUsers.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+var updateCanvas = require('./helpers/updateCanvas.js');
+var fs = require('fs');
+var options = {
+  headers: {
+    Authorization: KEYS.canvas.api_key,
+    'User-Agent': 'Request-Promise'
+  }
+};
+
+new CronJob(CONFIG.time_interval.cron_twice_per_day, findMeltedCanvasUsersAndDeleteThemInWiW);
```
Comment (tongxiang) [inline jobs/scheduling/dropMeltedUsers.js]: This CronJob will not run. 

Check and see what params this requires: 

https://github.com/ncb000gt/node-cron#api

**Comment by jrauschenberg** [comment]
Answered all your comments. New code pushed up. Let me know what you think, especially about the CronJob params. Thanks!

**Comment by tongxiang** [comment]
👍 looks good, John! Good work adding tests.
